File: RemoveDuplicatesFromSortedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        
        if (!head)
            return head;
        
        // use pointer to pointer, avoid prev pointer
        ListNode** start = &head;
        ListNode* node = NULL;
        while (*start) {
            if ((*start)->next) {
                if ((*start)->val == (*start)->next->val) {
                    node = (*start)->next;
                    (*start)->next = (*start)->next->next;
                    free(node);
                }
                else {
                    start = &((*start)->next);
                }
                continue;
            }
            break;
        }
        
        return head;
        
    }
};

File: jiuzhang/remove-duplicates-from-sorted-list/remove-duplicates-from-sorted-list.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""



"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        delflag = 1
        flag = 1
        p = head
        while(p != None and p.next != None):
            if p.val != p.next.val:
                flag = 1
                p = p.next
            elif flag < delflag:
                flag += 1;
                p = p.next
            else:
                p.next = p.next.next
        return head

